File: android/src/main/kotlin/com/dns_technologies/mlkit_scanner/commands/base/ScannerCommand.kt
package com.dns_technologies.mlkit_scanner.commands.base

import com.dns_technologies.mlkit_scanner.models.ScannerSession
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result

/** Base command abstraction for handling a single Dart->native scanner command. */
internal abstract class ScannerCommand(
    scannerSessionProvider: () -> ScannerSession?,
) : BaseScannerCommand(scannerSessionProvider) {
    /** Executes command body with shared error handling. */
    fun execute(call: MethodCall, result: Result) {
        try {
            executeCommand(call, result)
        } catch (error: Exception) {
            reportError(result, error)
        }
    }

    /** Command-specific implementation. */
    protected abstract fun executeCommand(
        call: MethodCall,
        result: Result,
    )
}

File: ios/Classes/Commands/Base/ScannerCommand.swift
import Flutter

/// Base command abstraction with shared synchronous error handling.
class ScannerCommand: BaseScannerCommand {
    /// Executes a command and reports any thrown validation or scanner error.
    final func execute(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        do {
            try executeCommand(call, result: result)
        } catch {
            reportError(result, error: error)
        }
    }

    /// Performs subclass-specific command work after common error handling.
    func executeCommand(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) throws {
        preconditionFailure("ScannerCommand subclasses must implement executeCommand")
    }
}
